File: BeFit iOS/FoodObjectsViewController.m
//
//  FoodObjectsViewController.m
//  BeFit iOS
//
//  Created by Satinder Singh on 2/17/17.
//  Copyright © 2017 Jon Brown. All rights reserved.
//

#import "FoodObjectsViewController.h"

@interface FoodObjectsViewController ()

@end

@implementation FoodObjectsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
    self.title =  [self.foodListObject valueForKey:@"name"] ;
   
    
    if ([[self.foodListObject valueForKey:@"name"] isEqualToString:@"User Food Library"])
    {
        _btnAdd.enabled = FALSE ;
        [_btnAdd setTintColor:[UIColor clearColor ]] ;
    }
    
    
}
-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:YES];
    
    isForEditing = FALSE ;
    items = [self.foodListObject valueForKey:@"foods"] ;
    foodObjectArray = [NSMutableArray arrayWithArray:[items allObjects]];
    [self.tableView reloadData];
}
- (NSManagedObjectContext *)managedObjectContext
{
    NSManagedObjectContext *context = nil;
    id delegate = [[UIApplication sharedApplication] delegate];
    if ([delegate performSelector:@selector(managedObjectContext)]) {
        context = [delegate managedObjectContext];
    }
    return context;
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {

    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {

    return foodObjectArray.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{

    // Configure the cell...
    static NSString *cellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:cellIdentifier];
    }
    
    cell.accessoryView = nil;
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    Food* food = [foodObjectArray objectAtIndex:indexPath.row];
    cell.textLabel.text = food.name ;
    
    return cell;
}



// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}

-(void)closeAlertview
{
    
    [self dismissViewControllerAnimated:YES completion:nil];
}

// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSManagedObjectContext *context = [self managedObjectContext];
    
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete object from database
        
        if ([[self.foodListObject valueForKey:@"name"] isEqualToString:@"User Food Library"])
        {
            selectedIndexPath = indexPath ;
            
            UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Befit" message:@"Food item deleted from User Food Library will remove it from all other libraries. Would you like to continue?" preferredStyle:UIAlertControllerStyleAlert];
            
            [alertController addAction:[UIAlertAction actionWithTitle:@"Yes" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
                
                NSManagedObjectContext *context = [self managedObjectContext];
                Food* foodObject = [foodObjectArray objectAtIndex:selectedIndexPath.row] ;
                NSString* foodName = foodObject.name ;
                foodObject = [AppDelegate CheckForDuplicateFoodItem:foodName][0];
                
                if (foodObject.userDefined.boolValue == TRUE)
                {
                    [foodObjectArray removeObjectAtIndex:selectedIndexPath.row];
                    
                    NSSet *distinctSet = [NSSet setWithArray:foodObjectArray];
                    
                    [self.foodListObject setValue:distinctSet forKey:@"foods"];
                    
                    [context deleteObject:foodObject];
                }
                else
                {
                    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Befit" message:@"Default food item cannot be deleted directly from this library" preferredStyle:UIAlertControllerStyleAlert];
                    
                    UIAlertAction* ok = [UIAlertAction actionWithTitle:@"Okay" style:UIAlertActionStyleDefault handler:nil];
                    [alertController addAction:ok];
                    
                    [self presentViewController:alertController animated:YES completion:nil];
                    
                    //            [[[UIAlertView alloc] initWithTitle:@"Befit" message:@"Default food item cannot be deleted directly from this library" delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles: nil]show];
                    return ;
                }
                NSError *error = nil;
                if (![context save:&error]) {
                    NSLog(@"Can't Delete! %@ %@", error, [error localizedDescription]);
                    return;
                }
                
                // Remove device from table view
                
                [self.tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:selectedIndexPath] withRowAnimation:UITableViewRowAnimationFade];
            }]];
            
            [alertController addAction:[UIAlertAction actionWithTitle:@"No" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
                [self closeAlertview];
            }]];
            
            dispatch_async(dispatch_get_main_queue(), ^ {
                [self presentViewController:alertController animated:YES completion:nil];
            });

            
            
           // [[[UIAlertView alloc] initWithTitle:@"Befit" message:@"Food item deleted from User Food Library will remove it from all other libraries. Would you like to continue?" delegate:self cancelButtonTitle:@"No" otherButtonTitles:@"Yes", nil]show];
           
            
        }
        else
        {
            [foodObjectArray removeObjectAtIndex:indexPath.row];
            
            NSSet *distinctSet = [NSSet setWithArray:foodObjectArray];
            
            [self.foodListObject setValue:distinctSet forKey:@"foods"];
            //        [context deleteObject: [foodObjectArray objectAtIndex:indexPath.row]];
            
            NSError *error = nil;
            if (![context save:&error]) {
                NSLog(@"Can't Delete! %@ %@", error, [error localizedDescription]);
                return;
            }
            
            // Remove device from table view
            
            [self.tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationFade];
            
//            Food* foodObject = [foodObjectArray objectAtIndex:indexPath.row] ;
//            
//            NSString* foodName = foodObject.name ;
//            
//            [foodObjectArray removeObjectAtIndex:indexPath.row];
//            
//            NSSet *distinctSet = [NSSet setWithArray:foodObjectArray];
//            
//            [self.foodListObject setValue:distinctSet forKey:@"foods"];
//            //        [context deleteObject: [foodObjectArray objectAtIndex:indexPath.row]];
//            
//            foodObject = [AppDelegate CheckForDuplicateFoodItem:foodName][0];
//            
//            NSSet* newFoodSet = foodObject.foodListsBelongTo ;
//            
//            NSArray* newFoodListArray = [NSArray arrayWithArray:[newFoodSet allObjects]];
//            
//            BOOL isItemFound = FALSE ;
//            FoodList* defaultLibrary ;
//            
//            for (int iCount = 0 ; iCount < newFoodListArray.count; iCount++)
//            {
//                FoodList* new_list = [newFoodListArray objectAtIndex:iCount] ;
//                if ([new_list.name isEqualToString:@"User Food Library"])
//                {
//                    isItemFound = true ;
//                }
//                else if ([new_list.name isEqualToString:@"Food Library"])
//                {
//                    isItemFound = true ;
//                    defaultLibrary = new_list ;
//                }
//            }
//            if (newFoodListArray.count == 1 && isItemFound == TRUE)
//            {
//                foodObject.foodListsBelongTo = [NSSet new] ;
//            }
//            else if ((defaultLibrary && isItemFound == TRUE) && newFoodListArray.count == 2)
//            {
//                foodObject.foodListsBelongTo = [NSSet setWithObject:defaultLibrary] ;
//            }
//            
//            NSError *error = nil;
//            if (![context save:&error]) {
//                NSLog(@"Can't Delete! %@ %@", error, [error localizedDescription]);
//                return;
//            }
//            
//            // Remove device from table view
//            
//            [self.tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationFade];
            
        }
       
       
        
        
    }
}
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    Food* food = [foodObjectArray objectAtIndex:indexPath.row];
    
    if (food.userDefined.boolValue)
    {
        isForEditing = TRUE ;
        [self performSegueWithIdentifier:@"segue_edit" sender:self];
    }
    else
    {
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Befit" message:@"Default food item cannot be edited" preferredStyle:UIAlertControllerStyleAlert];
        
        UIAlertAction* ok = [UIAlertAction actionWithTitle:@"Okay" style:UIAlertActionStyleDefault handler:nil];
        [alertController addAction:ok];
        
        [self presentViewController:alertController animated:YES completion:nil];
        
//        [[[UIAlertView alloc] initWithTitle:@"Befit" message:@"Default food item cannot be edited" delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles: nil]show];
    }
    
    
}

//- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex
//{
//    if (buttonIndex == 1)
//    {
//        
//        
//    }
//}

- (IBAction)AddNewFood:(id)sender
{
    [self performSegueWithIdentifier:@"segue_edit" sender:self];
}
#pragma mark prepareForSegue Functions

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"segue_edit"])
    {
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        AddFoodViewController *destViewController = segue.destinationViewController;
        if (foodObjectArray.count > 0)
        {
            destViewController.foodListObject = [foodObjectArray objectAtIndex:indexPath.row];
        }
        
        destViewController.isForEditing = isForEditing ;
        
    }
    
}

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/


@end

File: BeFit iOS/ViewControllers/FoodObjectsViewController.swift
//
//  FoodObjectsViewController.swift
//  BeFitTracker
//
//  Created by Andrej Slegl on 10/25/19.
//  Copyright © 2019 Jon Brown. All rights reserved.
//

import Foundation
import UIKit
import CoreData

class FoodObjectsViewController: UITableViewController {
    //MARK: Properties
    @IBOutlet weak var btnAdd: UIBarButtonItem!
    
    private var foodObjectArray: [Food] = []
    
    var foodListObject: FoodList?
    
    private var managedObjectContext: NSManagedObjectContext? {
        return (UIApplication.shared.delegate as? AppDelegate)?.managedObjectContext
    }
    
    //MARK: - Init
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Uncomment the following line to preserve selection between presentations.
        // self.clearsSelectionOnViewWillAppear = NO;
        
        // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
        // self.navigationItem.rightBarButtonItem = self.editButtonItem;
        title = foodListObject?.name
        
        if foodListObject?.name == "User Food Library" {
            btnAdd.isEnabled = false
            btnAdd.tintColor = .clear
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        foodObjectArray = (foodListObject?.foods?.allObjects as? [Food]) ?? []
        tableView.reloadData()
    }
    
    //MARK: - Private
    
    private func closeAlertview() {
        dismiss(animated: true, completion: nil)
    }
    
    private func deleteFoodItem(at indexPath: IndexPath) {
        guard let context = managedObjectContext else {
            assertionFailure()
            return
        }
        
        let foodName = foodObjectArray[indexPath.row].name
        
        guard let foodObject = AppDelegate.check(forDuplicateFoodItem: foodName ?? "").first else {
            assertionFailure()
            return
        }
        
        if foodObject.userDefined?.intValue == 1 {
            foodObjectArray.remove(at: indexPath.row)
            foodListObject?.removeFromFoods(foodObject)
            
            context.delete(foodObject)
        } else {
            let alertController = UIAlertController(title: "Befit", message: "Default food item cannot be deleted directly from this library", preferredStyle: .alert)
            alertController.addAction(UIAlertAction(title: "Okay", style: .default, handler: nil))
            
            present(alertController, animated: true)
            
            //            [[[UIAlertView alloc] initWithTitle:@"Befit" message:@"Default food item cannot be deleted directly from this library" delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles: nil]show];
            return
        }
        
        // Remove device from table view
        tableView.deleteRows(at: [indexPath], with: .fade)
        
        do {
            try context.save()
        } catch let error {
            print("Can't Delete!", error.localizedDescription)
            return
        }
    }
    
    //MARK: - Table view data source
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return foodObjectArray.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        
        cell.accessoryView = nil
        cell.accessoryType = .disclosureIndicator
        let food = foodObjectArray[indexPath.row]
        cell.textLabel?.text = food.name
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        // Return NO if you do not want the specified item to be editable.
        return true
    }
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        guard let context = managedObjectContext else {
            assertionFailure()
            return
        }
        
        if editingStyle == .delete {
            // Delete object from database
            
            if foodListObject?.name == "User Food Library" {
                let alertController = UIAlertController(title: "Befit", message: "Food item deleted from User Food Library will remove it from all other libraries. Would you like to continue?", preferredStyle: .alert)
                alertController.addAction(UIAlertAction(title: "Yes", style: .destructive) { [weak self] _ in
                    self?.deleteFoodItem(at: indexPath)
                })
                
                alertController.addAction(UIAlertAction(title: "No", style: .default, handler: nil))
                
                DispatchQueue.main.async {
                    self.present(alertController, animated: true)
                }
                
                // [[[UIAlertView alloc] initWithTitle:@"Befit" message:@"Food item deleted from User Food Library will remove it from all other libraries. Would you like to continue?" delegate:self cancelButtonTitle:@"No" otherButtonTitles:@"Yes", nil]show];
            } else {
                let food = foodObjectArray.remove(at: indexPath.row)
                foodListObject?.removeFromFoods(food)
                
                // Remove device from table view
                tableView.deleteRows(at: [indexPath], with: .fade)
                
                do {
                    try context.save()
                } catch let error {
                    print("Can't Delete!", error.localizedDescription)
                    return
                }
            }
        }
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let food = foodObjectArray[indexPath.row]
        
        if food.userDefined?.intValue == 1 {
            performSegue(withIdentifier: "segue_edit", sender: food)
        } else {
            let alertController = UIAlertController(title: "Befit", message: "Default food item cannot be edited", preferredStyle: .alert)
            alertController.addAction(UIAlertAction(title: "Okay", style: .default, handler: nil))
            
            present(alertController, animated: true)
            
            //        [[[UIAlertView alloc] initWithTitle:@"Befit" message:@"Default food item cannot be edited" delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles: nil]show];
        }
    }
    
    @IBAction private func AddNewFood(_ sender: AnyObject) {
        performSegue(withIdentifier: "segue_edit", sender: self)
    }
    
    //MARK: - prepareForSegue Functions
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        switch segue.destination {
        case let addFoodViewController as AddFoodViewController:
            let food = sender as? Food
            addFoodViewController.foodListObject = food
            addFoodViewController.selectedListArray = [foodListObject].compactMap { $0 }
            addFoodViewController.isForEditing = food?.userDefined?.intValue == 1
            
        default:
            break
        }
        
        super.prepare(for: segue, sender: sender)
    }
    
    /*
     // Override to support rearranging the table view.
     - (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
     }
     */
    
    /*
     // Override to support conditional rearranging of the table view.
     - (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
     // Return NO if you do not want the item to be re-orderable.
     return YES;
     }
     */
}
